js/trickleButtonModel.js

```diff
@@ -0,0 +1,186 @@
+   */
+  calculateButtonState(isButtonDisableForced = false) {
+    if (!this.isEnabled()) {
+      this.set({
+        _isButtonVisible: false,
+        _isButtonDisabled: true
+      });
+      return;
+    };
+
+    const isTrickleKilled = controller.isKilled;
+
+    const isStepUnlocked = this.isStepUnlocked() || isTrickleKilled;
+    const isFinished = this.isFinished() || isTrickleKilled;
+    const trickleConfig = getModelConfig(this);
+
+    const isButtonVisibleBeforeCompletion = (trickleConfig._button._styleBeforeCompletion !== 'hidden');
+    // Force button to be hidden after completion if _isFullWidth, otherwise absolutely positioned buttons will stack
+    const isButtonVisibleAfterCompletion = (trickleConfig._button._styleAfterClick !== 'hidden') && !trickleConfig._button._isFullWidth;
+
+    const isStepLockingCompletionRequired = this.isStepLockingCompletionRequired();
+
+    const isNoCompletionRequiredAndLockedVisible = (!isStepLockingCompletionRequired && !isFinished && isButtonVisibleBeforeCompletion);
+    const isNoCompletionRequiredAndUnlockedVisible = (!isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
+    const isNoCompletionRequiredAndFinishedVisible = (!isStepLockingCompletionRequired && isFinished && isButtonVisibleAfterCompletion);
+    const isStepLockedAndVisibleBeforeCompletion = (isStepLockingCompletionRequired && !isStepUnlocked && isButtonVisibleBeforeCompletion);
+    const isFinishedAndVisibleAfterCompletion = (isStepLockingCompletionRequired && isFinished && isButtonVisibleAfterCompletion);
+    const isStepUnlockedAndButtonIncomplete = (isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
+
+    const isButtonVisible = isNoCompletionRequiredAndLockedVisible ||
+      isNoCompletionRequiredAndUnlockedVisible ||
+      isNoCompletionRequiredAndFinishedVisible ||
+      isStepLockedAndVisibleBeforeCompletion ||
+      isFinishedAndVisibleAfterCompletion ||
+      isStepUnlockedAndButtonIncomplete;
+
+    const isButtonEnabledBeforeCompletion = (trickleConfig._button._styleBeforeCompletion !== 'disabled');
+    const isButtonEnabledAfterCompletion = (trickleConfig._button._styleAfterClick !== 'disabled');
+
+    const isNoCompletionRequiredAndLockedEnabled = (!isStepLockingCompletionRequired && !isFinished && isButtonEnabledBeforeCompletion);
+    const isNoCompletionRequiredAndUnlockedEnabled = (!isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
+    const isNoCompletionRequiredAndFinishedEnabled = (!isStepLockingCompletionRequired && isFinished && isButtonEnabledAfterCompletion);
+    const isStepUnlockedAndButtonIncompleteWithoutButtonDisabledForced = (isStepLockingCompletionRequired && isStepUnlockedAndButtonIncomplete && !isButtonDisableForced);
+    const isFinishedAndEnabledAfterCompletion = (isStepLockingCompletionRequired && isFinished && isButtonEnabledAfterCompletion);
```

Comment:
`isNoCompletionRequiredAndLockedEnabled`, `isNoCompletionRequiredAndUnlockedEnabled`, `isNoCompletionRequiredAndFinishedEnabled`, `isFinishedAndEnabledAfterCompletion` are very similar to 
`isNoCompletionRequiredAndLockedVisible`,  `isNoCompletionRequiredAndUnlockedVisible`, `isNoCompletionRequiredAndFinishedVisible` and  `isFinishedAndVisibleAfterCompletion`.

It might be worth adding a new function to handle these conditions, e.g.:


```getButtonState(isVisibleButtonState) {
  const isTrickleKilled = controller.isKilled;

  const isStepUnlocked = this.isStepUnlocked() || isTrickleKilled;
  const isFinished = this.isFinished() || isTrickleKilled;
  const trickleConfig = getModelConfig(this);

  const isButtonVisibleBeforeCompletion = (trickleConfig._button._styleBeforeCompletion !== 'hidden');
  // Force button to be hidden after completion if _isFullWidth, otherwise absolutely positioned buttons will stack
  const isButtonVisibleAfterCompletion = (trickleConfig._button._styleAfterClick !== 'hidden') && !trickleConfig._button._isFullWidth;
  const isButtonEnabledBeforeCompletion = (trickleConfig._button._styleBeforeCompletion !== 'disabled');
  const isButtonEnabledAfterCompletion = (trickleConfig._button._styleAfterClick !== 'disabled');
  const conditionBeforeCompletion = (isVisibleButtonState ? isButtonVisibleBeforeCompletion : isButtonEnabledBeforeCompletion);
  const conditionAfterCompletion = (isVisibleButtonState ? isButtonVisibleAfterCompletion : isButtonEnabledAfterCompletion);

  const isNoCompletionRequiredAndLocked = (!isStepLockingCompletionRequired && !isFinished && conditionBeforeCompletion);
  const isNoCompletionRequiredAndUnlocked = (!isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
  const isNoCompletionRequiredAndFinished = (!isStepLockingCompletionRequired && isFinished && conditionAfterCompletion);
  const isFinishedAfterCompletion = (isStepLockingCompletionRequired && isFinished && conditionAfterCompletion);
  const isStepUnlockedAndButtonIncomplete = (isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
  
  
  if (isVisibleButtonState) {
  const isStepLockedAndVisibleBeforeCompletion = (isStepLockingCompletionRequired && !isStepUnlocked && isButtonVisibleBeforeCompletion);
  const isStepUnlockedAndButtonIncomplete = (isStepLockingCompletionRequired && isStepUnlocked && !isFinished);
  return isNoCompletionRequiredAndLockedVisible ||
    isNoCompletionRequiredAndUnlocked ||
    isNoCompletionRequiredAndFinished ||
    isStepLockedAndVisibleBeforeCompletion ||
    isFinishedAfterCompletion ||
    isStepUnlockedAndButtonIncomplete;
  } else {
    const isStepUnlockedAndButtonIncompleteWithoutButtonDisabledForced = (isStepLockingCompletionRequired && isStepUnlockedAndButtonIncomplete && !isButtonDisableForced);
    
    return isNoCompletionRequiredAndLocked ||
    isNoCompletionRequiredAndUnlocked ||
    isNoCompletionRequiredAndFinished ||
    isStepUnlockedAndButtonIncompleteWithoutButtonDisabledForced ||
    isFinishedAfterCompletion ||
    false;
  }
}```

Reply:
I'm not entirely sure it'll increase the clarity splitting it like that. It might be better to have a function a piece and go for return earlies. But it's a complicated thing to spend time on to achieve a modicum more clarity at the moment.